### DISCUSSION TITLE
RE: Issue #34 - `pip install -e . equivalent?`

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [python-poetry/poetry#5276] [General]
Beginning this discussion here to continue questions and answers regarding issue #34.

**Comment by adam-grant-hendry** [comment]
@NeilGirdhar In continuation of our discussion, let's assume I wish to add `matplotlib` as a dependency to my project. Let's also assume I need to edit it to meet my user's requirements, but it wouldn't make sense for the `matplotlib` developers to add the edit in a PR because it wouldn't serve other `matplotlib` users. Hence, I would need to install this in development mode.

If I run

```bash
poetry run py -m pip install -e git+https://github.com/matplotlib/matplotlib.git#egg=matplotlib
```

this installs the package in development mode (under `.venv/src/`), but doesn't add `matplotlib` to my `pyproject.toml` (at least with `poetry 1.1`).

What should I put in my `pyproject.toml` to track it? This?:

```python
[too.poetry]
# ...
packages = [
    { include = "my_package" },
    { include = "matplotlib", from = ".venv/src" },
]
```

**Comment by NeilGirdhar** [comment]
So your user requires a modified version of matplotlib?   Why not fork matplotlib into a  new github repo, and then depend on that repo
```
poetry add my_matplotlib --git git://github.com/myorganization/matplotlib.git
```

**Reply by adam-grant-hendry** [reply]
@NeilGirdhar This is a theoretical example. No, I do not need a forked version of `matplotlib`.

If that is your answer, then that is what the `git submodule add` command is for. Please read [my original question](https://github.com/python-poetry/poetry/issues/34#issuecomment-1058767370).

**Reply by NeilGirdhar** [reply]
I'm trying to understand your workflow.  In the workflow you describe above, this is how I would do it.  How does this solution not satisfy your workflow?

Incidentally, aren't you getting tired of downvoting people's attempts at helping you?

**Reply by adam-grant-hendry** [reply]
I will always downvote an answer that is not helpful, no I am not tired of it. Voting gives other readers an idea of what users found helpful. I downvoted it because my original question contained this solution in the form of `git submodule add`. This tells me you didn't read my question clearly, which caused the issue to be litered with superfluous discussion, and consequently further frustrate others coming to the issue trying to find an answer.

**Reply by finswimmer** [reply]
Hello @adam-grant-hendry,

in your original question you was asking for best practices. For the case you need a modified version of another project, I agree with @NeilGirdhar: Fork this project, modify it and add it as a git dependency.

Until now I cannot see your use case for editable installs. While them see quite practical at a first glance, they made more problems than they solve. There are only a few cases - maybe just one case - where they are the solution of choice: You have a monorepo with multiple projects that depend on each other and even after deployment the paths hasn't change and you don't plan to build wheels or sdist packages.

Being able to work on multiple projects, that depend on each other, at the same time, is another often mentioned use case. That's what I was uses earlier as well. But to my experiences this makes lot of headache as well, when working in a team. If these project are really that tight coupled, that one cannot wait until a function is implemented and deployed the software architecture might be not optimal. Instead start developing these projects as subpackages and if you feel later this is something that is needed by other, split them out into its own package.

**Reply by adam-grant-hendry** [reply]
@finswimmer 

Thank you, and I agree, both with you and @NeilGirdhar: forking is the right way to go.

**Comment by arogozhnikov** [comment]
I find editable installs a critical feature. So I've made a test what happens if project in poetry-managed, but installed with `pip install -e .` and it seems to work.

How does it work and should I consider that supported? Is there any benefit of using that over pip + pyproject.toml?

**Reply by jonananas** [reply]
Works for me as well. Pip supports editable installs from pyproject.toml files since [21.3](https://pip.pypa.io/en/stable/news/#v21-3)

**Reply by arogozhnikov** [reply]
Thanks for pointing @jonananas 

My main concerns are if this combination pip + poetry + editable installs is somewhat coincidental and may be broken (poetry AFAIK does not support editable installs)

**Reply by arogozhnikov** [reply]
ok, now I got it. That doesn't actually work, it only "works" if you have setup.py, in this case pip sees that poetry does not make editable installs, and runs installation instead of relying on poetry.

Here is what I have after deleting setup.py and trying to install with `pip install -e .`:
```
ERROR: Project file:///path/to/project has a 'pyproject.toml' and its build backend is missing the 'build_editable' hook. Since it does not have a 'setup.py' nor a 'setup.cfg', it cannot be installed in editable mode. Consider using a build backend that supports PEP 660.
```

**Reply by NeilGirdhar** [reply]
@arogozhnikov You are mistaken. You can do editable installs with a poetry project f you use the poetry build backend from master.  You can search the other threads for how to do this.

**Reply by jonananas** [reply]
I am using poetry 1.1.13 and pip 22.0.4, `pip3 install -e .` without `setup.py` works for me. For example I can do:
```bash
poetry new foo
pip3 install -e foo
python3 
>>> import foo
>>> print(foo.__version__)
```

**Reply by arogozhnikov** [reply]
My previous test was in ubuntu docker and both pip and poetry were updated to the latest available versions from pypi. There could be some mistake in my setup, but I can't what kind of mistake.

**Reply by arogozhnikov** [reply]
From looking at the error, pip I used already had support for editable installs + poetry hook looked like `requires = ["poetry>=1.1.0"]`, so ... 🤷

**Reply by NeilGirdhar** [reply]
Did you set the backend in your pyproject.toml?  `build-backend = 'poetry.core.masonry.api'`  There's a way to grab the one off github, but you'll have to find that.

**Reply by arogozhnikov** [reply]
I did

**Reply by jonananas** [reply]
@arogozhnikov I think you may be using the default pip 20.0.2 that comes with Ubuntu instead of pip >= 21.3
That's what happens if I do
```bash
apt-get update
apt-get install python3
apt-get install python3-pip
pip3 install poetry
poetry new foo
pip3 install -e foo
```
What does `pip3 --version` report?

**Reply by abn** [reply]
Curious, why not just use:

```sh
# poetry <= 1.1.13
poetry install --no-dev

# poetry >= 1.2.0a2
poetry install --only default
```

The root project is installed as an editable project.

**Reply by phaseOne** [reply]
Looks like you need [pip >= 21.3.1](https://pip.pypa.io/en/stable/news/#v21-3-1), which is the first stable version that supports [PEP 660](https://peps.python.org/pep-0660/), which introduces support for editable installs for pyproject.toml based builds. Sadly this version of pip is not available on Raspberry Pi OS buster, so I'll have to upgrade to bullseye.

**Comment by jonananas** [comment]
I believe it's the beauty of using a standard, i.e. that the fact that both pip and poetry supports `pyproject.toml` is the reason it works. But I am also not 100% certain.

**Comment by fredrikaverpil** [comment]
@adam-grant-hendry I would clone down the matplotlib repo locally. Then I would depend on it (in editable mode) via the [`path` directive](https://python-poetry.org/docs/master/dependency-specification/#path-dependencies) in my project:

```bash
git clone https://github.com/matplotlib/matplotlib.git ../matplotlib
```

```toml
[tool.poetry.dependencies]
matplotlib = {path = "../matplotlib", develop = true}
```

Note: for matplotlib, you need to also install setuptool-scm, as that is required for building from source:

```bash
poetry add --dev setuptools-scm
```

After a `poetry install` in your project, your project is using the code from your local `../matplotlib` folder. Edit the files in `../matplotlib` as you wish. Normally, you wouldn't have to re-run `poetry install` for making changes take effect, but with matplotlib I'm not sure how this works...
